### PR TITLE
Fix sof_8336 gpio problems

### PR DESCRIPTION
**Pull request by yangxiaohua2009**
Currently the driver uses quirks to define gpios, which means when the machine has two gpios, the user won't get sound from headphone by default. The patch fix the following problems:
1. Use two gpios by default
2. remove redundant quirks
3. Invert gpio status so that true means on and false means off.

Signed-off-by: Zhu Ning <zhuning@everest-semi.com>

### CODE REVIEW

**Comment by yangxiaohua2009** [comment]
Basically using two GPIOs by default is fine and the system won't crash. See [gpio.zip].(https://github.com/thesofproject/linux/files/10138979/gpio.zip)
Also, the first GPIO is always used for Speaker, and the second used for Headphone so some quirk can be removed.

**Comment by yangxiaohua2009** [comment]
Problems remains: how should the Headphone GPIO be controled? Should another power event, i.e., headphone_power_event be used? @plbossart

File: sound/soc/intel/boards/sof_es8336.c [hunk 1]
```diff
@@ -72,40 +69,21 @@ struct sof_hdmi_pcm {
...
 	{ "headphone-enable-gpios", &enable_gpio1, 1, ACPI_GPIO_QUIRK_ONLY_GPIOIO },
 	{ }
 };
 
-static const struct acpi_gpio_mapping acpi_enable_both_gpios_rev_order[] = {
-	{ "speakers-enable-gpios", &enable_gpio1, 1, ACPI_GPIO_QUIRK_ONLY_GPIOIO },
-	{ "headphone-enable-gpios", &enable_gpio0, 1, ACPI_GPIO_QUIRK_ONLY_GPIOIO },
-	{ }
-};
-
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:98]: why is this removed? this was added for a reason...

File: sound/soc/intel/boards/sof_es8336.c [hunk 2]
```diff
@@ -722,36 +692,28 @@ static int sof_es8336_probe(struct platform_device *pdev)
...
+	priv->gpio_speakers = gpiod_get_optional(codec_dev, "speakers-enable", GPIOD_OUT_HIGH);
 	if (IS_ERR(priv->gpio_speakers)) {
 		ret = dev_err_probe(dev, PTR_ERR(priv->gpio_speakers),
 				    "could not get speakers-enable GPIO\n");
 		goto err_put_codec;
 	}
+	gpiod_set_value_cansleep(priv->gpio_speakers, false);
 
-	priv->gpio_headphone = gpiod_get_optional(codec_dev, "headphone-enable", GPIOD_OUT_LOW);
+	priv->gpio_headphone = gpiod_get_optional(codec_dev, "headphone-enable", GPIOD_OUT_HIGH);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:710]: is this intentional? why would the GPIO level change?

File: sound/soc/intel/boards/sof_es8336.c [hunk 3]
```diff
@@ -802,7 +764,6 @@ static const struct platform_device_id board_ids[] = {
 					SOF_HDMI_CAPTURE_1_SSP(0) |
 					SOF_HDMI_CAPTURE_2_SSP(2) |
 					SOF_SSP_HDMI_CAPTURE_PRESENT |
-					SOF_ES8336_SPEAKERS_EN_GPIO1_QUIRK |
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:805]: if gpio1 was used, why did you remove this?

**Review by plbossart — CHANGES_REQUESTED** [review]
very very difficult to review this, this PR removes things that were added to deal with GPIO inversions and this also changes the GPIO level?

File: sound/soc/intel/boards/sof_es8336.c (hunk 1 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:98]: When two GPIOs are used, Speaker GPIO is always the first and HP GPIO the second. The user may need a quirk to invert the GPIO level.

File: sound/soc/intel/boards/sof_es8336.c (hunk 2 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:710]: In the past speaker_en false means on (high) and true means off (low) because the code is taken from bytcht driver. The patch inverts this.

File: sound/soc/intel/boards/sof_es8336.c (hunk 3 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:805]: On Windows GPIO0 is always the speaker GPIO. The quirk may be reused to invert the GPIO level. (The results are the same: either invert the level or invert the order.)

Comment (plbossart) [reply sound/soc/intel/boards/sof_es8336.c:805]: I can't follow this  "either invert the level or invert the order". How is this equivalent?

Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:805]: |   | GPIO1 | GPIO2 |
| - | - | - |
| Speaker on | high | low |
| Headphone on | low | high |

Inverting the high/low level or changing the GPIO order are basically the same.

File: sound/soc/intel/boards/sof_es8336.c [hunk 4]
```diff
@@ -119,8 +97,7 @@ static void pcm_pop_work_events(struct work_struct *work)
 
 	gpiod_set_value_cansleep(priv->gpio_speakers, priv->speaker_en);
 
-	if (quirk & SOF_ES8336_HEADPHONE_GPIO)
-		gpiod_set_value_cansleep(priv->gpio_headphone, priv->speaker_en);
+	gpiod_set_value_cansleep(priv->gpio_headphone, true);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:100]: why is the setting for headphone independent from the speaker enablement? Shouldn't the value be !priv->speaker_en?

File: sound/soc/intel/boards/sof_es8336.c [hunk 5]
```diff
@@ -158,10 +135,10 @@ static int sof_es8316_speaker_power_event(struct snd_soc_dapm_widget *w,
 	struct snd_soc_card *card = w->dapm->card;
 	struct sof_es8336_private *priv = snd_soc_card_get_drvdata(card);
 
-	if (priv->speaker_en == !SND_SOC_DAPM_EVENT_ON(event))
-		return 0;
+        dev_info(card->dev, "sof_es8316_speaker_power_event %d\n",SND_SOC_DAPM_EVENT_ON(event));
+        /* Redundant check, removed */
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:139]: why is it redundant? and if yes, it should be a separate patch to avoid confusing reviewers between different functionality and code cleanups.

File: sound/soc/intel/boards/sof_es8336.c [hunk 6]
```diff
@@ -158,10 +135,10 @@ static int sof_es8316_speaker_power_event(struct snd_soc_dapm_widget *w,
 	struct snd_soc_card *card = w->dapm->card;
 	struct sof_es8336_private *priv = snd_soc_card_get_drvdata(card);
 
-	if (priv->speaker_en == !SND_SOC_DAPM_EVENT_ON(event))
-		return 0;
+        dev_info(card->dev, "sof_es8316_speaker_power_event %d\n",SND_SOC_DAPM_EVENT_ON(event));
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:138]: dev_info is way too verbose.

File: sound/soc/intel/boards/sof_es8336.c [hunk 7]
```diff
@@ -302,6 +279,8 @@ static int sof_es8316_init(struct snd_soc_pcm_runtime *runtime)
 	snd_jack_set_key(priv->jack.jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
 
 	snd_soc_component_set_jack(codec, &priv->jack, NULL);
+	ret = snd_soc_dapm_get_pin_status(&card->dapm, "Speaker");
+	dev_info(card->dev, "Speaker status %d", ret);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:283]: why is this needed?

File: sound/soc/intel/boards/sof_es8336.c [hunk 8]
```diff
@@ -722,36 +692,28 @@ static int sof_es8336_probe(struct platform_device *pdev)
...
 
-	priv->gpio_headphone = gpiod_get_optional(codec_dev, "headphone-enable", GPIOD_OUT_LOW);
+	priv->gpio_headphone = gpiod_get_optional(codec_dev, "headphone-enable", GPIOD_OUT_HIGH);
 	if (IS_ERR(priv->gpio_headphone)) {
 		ret = dev_err_probe(dev, PTR_ERR(priv->gpio_headphone),
 				    "could not get headphone-enable GPIO\n");
 		goto err_put_codec;
 	}
-
+
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:716]: spurious line change

File: sound/soc/intel/boards/sof_es8336.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_es8336.c:710]: but this is going to break existing solutions, I really don't understand this change.

**Review by plbossart — CHANGES_REQUESTED** [review]
It still fundamentally do not understand how the levels of the GPIOs is determined. It seems that for speakers it was changed from LOW to HIGH, so it must have some effects on existing users.

Can we not use the _DSM information to determine what the active level is for GPIO0 and GPIO1?

**Comment by yangxiaohua2009** [comment]
> It still fundamentally do not understand how the levels of the GPIOs is determined. It seems that for speakers it was changed from LOW to HIGH, so it must have some effects on existing users.
> 
> Can we not use the _DSM information to determine what the active level is for GPIO0 and GPIO1?

Yes there are values defined for GPIO in the DSM, but I don't know how to get them in the machine driver.
```
/*
* TO DECLARE SPEAKER CONTROL LEVEL
* RETURN VALUE = 1, High level enable
*		 0,low level enable
*/
#define  SPK_CTL_IO_LEVEL 0x5B

/* TO DECLARE HEADPHONE SWITCH CONTROL LEVEL
 *	RETURN VALUE = 1, High level enable
 *		       0, low level enable
 */
#define HPSWITCH_CTL_IO_LEVEL 0x4B
```

File: sound/soc/intel/boards/sof_es8336.c (hunk 5 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:139]: This check will return 0 when first time the event is called. So the gpios are always the default value the first time sound is played. Should be a seperate patch.

File: sound/soc/intel/boards/sof_es8336.c (hunk 4 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:100]: |   | Speaker GPIO |  Headphone GPIO |
| - | - | - |
| Playing music Speaker on | high | low |
| Playing music Headphone on | low | high |
| Suspend | low | low |

Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:100]: When only one GPIO is used things are simple. The GPIO is high by default. Before playing music, the SPEAKER_ON dapm event is called and the GPIO set to high. When music stop, the GPIO is set to low so that there is no pop noise.

When HP GPIO is involved things become complicated. I want to do the same for HP GPIO but I don't know how. Maybe another [SND_SOC_DAPM_SUPPLY](https://elixir.bootlin.com/linux/v6.1-rc7/C/ident/SND_SOC_DAPM_SUPPLY) for headphone should be invloved or maybe the headphone status can be read by `snd_soc_dapm_get_pin_status(&card->dapm, "Headphone");` and the headphone can be turned on and off in [sof_8336_trigger](https://elixir.bootlin.com/linux/v6.1-rc7/C/ident/sof_8336_trigger)

**Comment by plbossart** [comment]
@yangxiaohua2009 I'll close this PR since we have no line-of-sight for any update or merge timeline. I have done enough on my side to try and help with this ES8336 support, I don't have time to help further.
